sanitize_html/lib/src/sane_html_validator.dart

```diff
@@ -217,35 +220,53 @@ class SaneHtmlValidator implements html.NodeValidator {
     @required this.allowClassName,
   });
 
-  @override
-  bool allowsAttribute(html.Element element, String attribute, String value) {
-    if (!_allowedElements.contains(element.tagName)) {
-      return false;
-    }
-    if (_alwaysAllowedAttributes.contains(attribute)) {
-      return true;
-    }
-    // Special validators for id and class on all elements
-    if (attribute == 'id') {
-      return allowElementId(element.id);
+  String sanitize(String htmlString) {
+    final root = html_parser.parseFragment(htmlString);
+    _sanitize(root);
+    return formatHtmlNode(root);
+  }
+
+  void _sanitize(Node node) {
+    if (node is Element) {
+      final tagName = node.localName.toUpperCase();
+      if (!_allowedElements.contains(tagName)) {
+        node.remove();
+        return;
+      }
+      node.attributes.removeWhere((k, v) {
+        final attrName = k.toString();
+        if (attrName == 'id') {
+          return allowElementId == null || !allowElementId(v);
+        }
+        if (attrName == 'class') {
+          if (allowClassName == null) return true;
+          node.classes.removeWhere((cn) => !allowClassName(cn));
+          return node.classes.isEmpty;
+        }
+        return !_isAttributeAllowed(tagName, attrName, v);
+      });
     }
-    if (attribute == 'class') {
-      return element.classes.every(allowClassName);
+    if (node.hasChildNodes()) {
+      for (int i = node.nodes.length - 1; i >= 0; i--) {
+        _sanitize(node.nodes[i]);
```

Comment:
Maybe leave a comment that we do this in reverse, because we could otherwise skip one, when a node is removed...